added test for Report Problem and added new functions in the InstrumentationUtility.kt

kbo001 opened the pull request:
## Summary
Automated the testcase for Report Problem in the menu Tab of the Mobile Test Plan:
https://docs.google.com/spreadsheets/d/1X9YAf_g1Cq9vZZdCkcBLnRSFrhZuk_vEyCSM3sD5QvM/edit#gid=1556040007
Also added functions  matchTypedText(), hideKeyboard(),verifyIntentText() and hardPressBack() in the  InstrumentationUtility.kt

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. Would be good to add screenshots/videos for any major user facing changes -->

## Safety Assurance

- [ ] If the PR is high risk, "High Risk" label is set
- [ ] I have confidence that this PR will not introduce a regression for the reasons below
- [ ] Do we need to enhance manual QA test coverage ? If yes, "QA Note" label is set correctly

### Automated test coverage

The testcase covers the submission of Report Problem form and that rotating the screen orientation does not effect the data present in the form.


### Safety story
Executed the test on local AVD and executed successfully

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LogSubmissionTest.kt:
```diff
@@ -42,4 +51,25 @@ class LogSubmissionTest: BaseTest() {
+        InstrumentationUtility.rotatePortrait()
+        InstrumentationUtility.hideKeyboard()
+        onView(withText("Submit Report")).perform(click())
+        InstrumentationUtility.stubIntentWithAction(Intent.ACTION_SEND)
```
we can remove this as it doesn't do anything here. This is only saying that if there is even an intent with action `Intent.ACTION_SEND`, return with the success result. It's not making sure that our intent actually had `Intent.ACTION_SEND`

kbo001 replied:
removed the step from the test.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LogSubmissionTest.kt:
```diff
@@ -42,4 +51,25 @@ class LogSubmissionTest: BaseTest() {
+        InstrumentationUtility.hideKeyboard()
+        onView(withText("Submit Report")).isPresent()
+        InstrumentationUtility.rotateLeft()
+        InstrumentationUtility.matchTypedText(R.id.ReportText01,"This is a test for Report Problem")
```
you should put the input text in a variable so that we can verify it later without repeating it.

kbo001 replied:
assigned the text to a variable and used the same.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/LogSubmissionTest.kt:
```diff
@@ -42,4 +51,25 @@ class LogSubmissionTest: BaseTest() {
+        InstrumentationUtility.hideKeyboard()
+        onView(withText("Submit Report")).perform(click())
+        InstrumentationUtility.stubIntentWithAction(Intent.ACTION_SEND)
+        InstrumentationUtility.verifyIntentText("This is a test for Report Problem")
```
this is not doing anything right now as you are not adding the intent callback generated by `verifyIntentText` to `IntentMonitorRegistry`. Also addition of callback needs to happen before we click on the submit button so that we can capture intent when it's fired on clicking submit. This is the code you need to have instead -  

````
        var intentcallback = verifyIntent(inputText)
        IntentMonitorRegistry.getInstance().addIntentCallback(intentcallback)
        onView(withText("Submit Report")).perform(click())
        IntentMonitorRegistry.getInstance().removeIntentCallback(intentcallback)
    }


    fun verifyIntent(inputText: String) = IntentCallback { intent ->
        var extraText = intent.extras!!.getString(Intent.EXTRA_TEXT)
        assertTrue(extraText!!.contains(inputText))
        assertTrue(intent.action.contentEquals(Intent.ACTION_SEND))
    }
````

kbo001 replied:
Renamed the earlier function that validated the callback to verifyIntentTextOnSubmitReport() and added all the steps in the same function. I tried to add the steps in the test file but was getting error of "inferred type" so added them in the utility file.

kbo001 replied:
Also removed InstrumentationUtility.stubIntentWithAction(Intent.ACTION_SEND)

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -283,6 +287,13 @@ object InstrumentationUtility {
         Espresso.closeSoftKeyboard()
     }
 
+    /**
```
Empty java doc, either add a description or remove the comments.

kbo001 replied:
descriptions added.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -283,6 +287,13 @@ object InstrumentationUtility {
+    /**
+     *
+     */
+    fun matchTypedText(@IdRes editTextId: Int, text: String){
```
great job on adding the function to our utility class, this is a general purpose function that can be used at numerous places and hence a perfect fit for this class.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -310,6 +321,21 @@ object InstrumentationUtility {
+        onView(ViewMatchers.isRoot()).perform(ViewActions.closeSoftKeyboard())
+    }
+
+    fun verifyIntentText(text : String){
```
this function is better suited in the test class itself instead of this utility class as I don't think there are much chances of this getting used by another test.

kbo001 replied:
removed the function and added the steps in the function

shubham1g5 reviewed (changes requested):
Great attempt with these tests, I left some comments on how to work the intent verification better. Feel free to clarify any questions you have regarding any of my comments.

shubham1g5 commented:
@damagatchi retest this please

dimagimon commented:
<!--
  0 Errors
  1 Warning: This PR does not contain any J...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR does not contain any JIRA issue keys in the PR title or commit messages (e.g. KEY-123)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -310,6 +318,18 @@ object InstrumentationUtility {
         }
     }
 
+    fun verifyIntentTextOnSubmitReport(text : String){
```
we should place this function inside `LogSubmissionTest` itself. We only add general purpose functions in this call which can be utilised across different tests and this function is very specific to the report problem test.

kbo001 replied:
I have added the steps inside the test itself and removed the function from here.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -310,6 +318,18 @@ object InstrumentationUtility {
+        var intentcallback = IntentCallback { intent ->
+            var extraText = intent.extras!!.getString(Intent.EXTRA_TEXT)
+            assertTrue(extraText!!.contains(text))
+        }
```
you can also add a check for intent action like `assertTrue(intent.action.contentEquals(Intent.ACTION_SEND))`

kbo001 replied:
I tried adding the assertion for checking the action but it kept failing as it returned a different intent (CHOOSER). So didn't add the same.

shubham1g5 commented:
@kbo001 Seems like there is an issue with browserstack device we are using due to which all our [tests are failing](https://app-automate.browserstack.com/dashboard/v2/builds/7cad7994a30a7f679aa6cf48f1359fb75d44c614/sessions/tests/0e28331e5b0fc0151489e08e543efafdbd7576174f69618b). I have written to their support regarding this issue and will update when I hear back from them.

shubham1g5 reviewed (approved):
Looks good to me! Though we will need to wait for Browserstack issue to be fixed for this to get run and tested over there.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -442,5 +446,53 @@ object InstrumentationUtility {
+     * A utility to verify Form Fields with their values
+     */
+
+    fun verifyFormCellAndValue(cellData: String, value: String): Boolean {
```
this is a great addition!

kbo001 replied:
Thank you!

shubham1g5 reviewed (changes requested):
Great effort again on the test! Really excited to see this turning out well.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+                .perform(ViewActions.click())
+            if (caseList[2] == "Yesterday") {
+                InstrumentationUtility.setDateTo(1, "Past")
+                Thread.sleep(1000)
```
without the sleep the test fails, without clicking the next button.

kbo001 replied:
added a comment regarding the sleep next to the step.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+
+            onView(ViewMatchers.withClassName(Matchers.endsWith("EditText")))
+                .perform(ViewActions.typeText((caseList[1])))
+            onView(ViewMatchers.withId(R.id.nav_btn_finish))
```
added a InstrumentationUtility function submitForm()

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+                .perform(ViewActions.typeText((caseList[0])))
+            onView(ViewMatchers.withId(R.id.nav_btn_next))
+                .perform(ViewActions.click())
+            if (caseList[2] == "Yesterday") {
+                InstrumentationUtility.setDateTo(1, "Past")
+                Thread.sleep(1000)
+                onView(ViewMatchers.withId(R.id.nav_btn_next))
+                    .perform(ViewActions.click())
+            } else if (caseList[2] == "Today") {
+                onView(ViewMatchers.withId(R.id.nav_btn_next))
+                    .perform(ViewActions.click())
+            }
```
made the changes

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+                .perform(ViewActions.click())
+            onView(ViewMatchers.withClassName(Matchers.endsWith("EditText")))
+                .perform(ViewActions.typeText((caseList[0])))
+            onView(ViewMatchers.withId(R.id.nav_btn_next))
```
used the utility as requested

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+    @Test
+    fun testCreateCases(){
+        for (caseList in listOfCases) {
+            onView(ViewMatchers.withText("Start"))
+                .perform(ViewActions.click())
+            onView(ViewMatchers.withText("Case List"))
+                .perform(ViewActions.click())
+            onView(ViewMatchers.withText("Register"))
+                .perform(ViewActions.click())
```
Used the utility in all the tests.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Test
+    fun testUpdateCase2(){
```
Merged both the testUpdateCase into 1 and also used loop in the testVerifyChanges test.

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,163 @@
+        InstrumentationUtility.clickListItem(R.id.screen_entity_select_list, 0)
+        Thread.sleep(2000)
+        assertTrue(InstrumentationUtility.verifyFormCellAndValue("Appointment Attended?","Seen by doctor"))
+        InstrumentationUtility.hardPressBack()
```
used the same.

kbo001 commented on app/instrumentation-tests/src/org/commcare/utils/InstrumentationUtility.kt:
```diff
@@ -442,5 +446,53 @@ object InstrumentationUtility {
         }
     }
+
+    fun setDateTo(days: Int, past_present_future : String){
```
moved the function to the testfile and made the changes as requested.

kbo001 commented:
@shubham1g5 I have made the changes that have been requested. 
1. removed setDateTo() from utility file and added in the test file,
2.  merged both update tests into 1, used loop in the testVerifyChanges, 
3.  added tearDown method to testfile, 
4.  replaced few of the teststeps with already existing utility methods
5. added annotation to execute the tests in order

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,162 @@
+     * Adding the teardown method so the tests doesnot fail after only thefirst execution, when executed as a whole.
+     */
+    @After
+    fun tearDown(){
```
the downside of having 3 different test methods below when we can have 1 is that after every test we will end up uninstalling and installing the app which can be a big factor of total test duration. A better approach would be - 

````
@Test
fun testConditionalEnumInCaseListTests(){
   createCases();
   updateCases();
   verifyCases();
}

fun createCases(){...}
fun updateCases(){...}
fun verifyCases(){...}
````

kbo001 replied:
merged all tests into 1.

kbo001 replied:
Also removed the tearDown method.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,162 @@
+    fun testCreateCases(){
+        for (caseList in listOfCases) {
+            InstrumentationUtility.openForm(0,0)
+            onView(ViewMatchers.withClassName(Matchers.endsWith("EditText")))
```
can abstract as `InstrumentationUtility.enterText()`

kbo001 replied:
one utility was already present with same name that required element id, so created another without ID argument.

shubham1g5 commented on app/instrumentation-tests/src/org/commcare/androidTests/ConditionalEnumInCaseListTests.kt:
```diff
@@ -0,0 +1,162 @@
+    fun testUpdateCases(){
+        for (caseList in listOfCases) {
+            InstrumentationUtility.openForm(0,1)
+            onView(ViewMatchers.withId(R.id.search_action_bar)).perform(ViewActions.click())
+            onView(ViewMatchers.withId(R.id.search_src_text)).perform(ViewActions.typeText(caseList[0]))
+            onView(ViewMatchers.withId(R.id.screen_entity_detail_list)).isPresent()
+            InstrumentationUtility.clickListItem(R.id.screen_entity_select_list, 0)
```
these 4 lines can be abstracted as `InstrumentationUtility.searchAndEnterCase()` and can be resued in verification as well.

kbo001 replied:
created the same.

shubham1g5 reviewed (changes requested):
looks great!

kbo001 commented on app/instrumentation-tests/src/org/commcare/androidTests/MenuBadgeTests.kt:
```diff
@@ -0,0 +1,162 @@
+@LargeTest
+@BrowserstackTests
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)  //executes the tests in order
+class ConditionalEnumInCaseListTests: BaseTest() {
```
removed the same.

shubham1g5 commented:
Browserstack's team got back to me saying that issue has been fixed. 
@damagatchi retest this please

shubham1g5 commented:
There is a unrelated test failure that's causing the build to fail, I have [pulled a fix](https://github.com/dimagi/commcare-android/pull/2565) for that. Once that's merged we should merge `master` in this branch to get the build running again.

shubham1g5 reviewed (approved):
Build passed. This is good to go @kbo001 🥳
